### PR TITLE
Buffer ptr null

### PR DESCRIPTION
**Pull request by mlaures**

### CODE REVIEW

File: testing/test_impl.hpp [hunk 1]
```diff
@@ -14,12 +14,59 @@
...
+    //test create, that it's null
+    auto s = buffer_ptr<T>();
+    EXPECT_FALSE(s) << "Null buffer pointer is not null";
+    
+    /* Use Counts cannot be used right now because the shared_ptr compiler 
+     * implementation and the stdlib definition tend to disagree with with 
+     * the actual counts should be for empty shared_ptrs
+     * https://stackoverflow.com/questions/48885252/c-sharedptr-use-count-for-nullptr
+     */
+    // UnexpectedEq(s.use_count(), 1, "reference count");
```
Comment (mpokorny) [inline testing/test_impl.hpp:27]: It's OK to omit this test; as your comment indicates, there is ambiguity surrounding the use count of an empty `shared_ptr`. I see that you already have a similar test for a non-empty `buffer_ptr`, and that should be good enough.

**Review by mpokorny — CHANGES_REQUESTED** [review]
Looks pretty good, mostly just minor suggestions. The only comment that I suggest you definitely address is adding the `std` namespace to `nullptr_t`.

File: buffer.hpp [hunk 2]
```diff
@@ -28,6 +28,12 @@ auto reuseable_buffer<T>::operator[](unsigned int i) const noexcept -> T& {
     return *(ptr + i);
 }
 
+template <typename T>
+buffer_ptr<T>::buffer_ptr() {
+    sp = std::shared_ptr<reuseable_buffer<T>>(nullptr);
```
[the remark this answers is not in the document]
Comment (mlaures) [reply buffer.hpp:33]: This is the default constructor for buffer_ptr so I can't be using `sp.reset()` yet

File: buffer.hpp [hunk 3]
```diff
@@ -58,8 +64,29 @@ auto buffer_ptr<T>::operator[](int i) const noexcept -> T&{
...
+
+/* TODO: Is this confusing? The buffer_ptr doesn't necessarily hold a nullptr, 
+ * it is just setting shared_ptr to empty. 
+ * 
+ * Side Note: As far as I can find 
+ * there is no real way for a shared_ptr to hold a null_ptr properly anyway, 
+ * they just tend to be empty.
+ */
+template <typename T>
+buffer_ptr<T>& buffer_ptr<T>::operator=(std::nullptr_t) noexcept {
```
[the remark this answers is not in the document]
Comment (mlaures) [reply buffer.hpp:80]: Thanks for pointing out, I added the reset and more assignment operators can be defined as needed.

**Review by mpokorny — APPROVED** [review]
LGTM

**Pull request merged**